src/gc.c

```diff
@@ -77,25 +78,29 @@ static void threadCallback(void* data) {
   }
 
   if (!ret) {
-    rm_free (task);
+    // The index was freed. There is no need to reschedule the task.
+    // We need to free the task and the GC.
+    RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC %p: Self-Terminating. Index was freed.", gc);
+    gc->callbacks.onTerm(gc->gcCtx);
+    if (bc) RedisModule_UnblockClient(bc, NULL);
+    rm_free(task);
+    rm_free(gc);
     goto end;
   }
 
   RedisModule_ThreadSafeContextLock(ctx);
-  gc->timerID = scheduleNext(task);
+  if (gc->timerID) {
+    gc->timerID = scheduleNext(task);
```

Comment:
Theoretically we only need to lock Redis here right? Not on the else.
Also does Redis promise that the timerID we get will never be zero?

Reply:
1. Right, but also for the condition check. Seems fine to me to perform the else section locked for simplicity (we should usually satisfy the condition)
2. We seems to behave this way on the GC-start function below as an error value, so I assumed it is

Reply:
https://github.com/RediSearch/RediSearch/pull/1154#discussion_r403673323

Reply:
from redis:

```
    uint64_t expiretime = ustime()+period*1000;
    uint64_t key;

    while(1) {
        key = htonu64(expiretime);
        if (raxFind(Timers, (unsigned char*)&key,sizeof(key)) == raxNotFound) {
            raxInsert(Timers,(unsigned char*)&key,sizeof(key),timer,NULL);
            break;
        } else {
            expiretime++;
        }
    }
```
So it seems very unlikely to get 0, at least the same as `UINT64_MAX`. WDYT?